test/core/util/tls_utils.cc

```diff
@@ -71,6 +72,68 @@ std::string GetFileContents(const char* path) {
   return credential;
 }
 
+absl::Status CheckPrivateKey(absl::string_view private_key) {
+  if (private_key.empty()) {
+    return absl::InvalidArgumentError("Private key string is empty.");
+  }
+  BIO* private_key_bio =
+      BIO_new_mem_buf(private_key.data(), private_key.size());
+  if (private_key_bio == nullptr) {
+    return absl::InvalidArgumentError(
+        "Conversion from private key string to BIO failed.");
+  }
+  EVP_PKEY* private_evp_pkey =
+      PEM_read_bio_PrivateKey(private_key_bio, nullptr, nullptr, nullptr);
+  if (private_evp_pkey == nullptr) {
+    BIO_free(private_key_bio);
+    return absl::InvalidArgumentError("Invalid private key string.");
+  }
+  int pkey_type = EVP_PKEY_id(private_evp_pkey);
+  bool is_key_type_defined = true;
+  switch (pkey_type) {
+    case EVP_PKEY_NONE:
+      is_key_type_defined = false;
+      break;
+    case EVP_PKEY_RSA:
+    case EVP_PKEY_RSA_PSS:
+      // The cases that lead here represent currently supported key types.
+      break;
+    default:
+      gpr_log(GPR_ERROR, "Key type currently not supported.");
+  }
+  BIO_free(private_key_bio);
+  EVP_PKEY_free(private_evp_pkey);
+  return is_key_type_defined
+             ? absl::OkStatus()
+             : absl::InvalidArgumentError("Undefined key type.");
+}
+
+absl::Status CheckCertChain(absl::string_view cert_chain) {
+  if (cert_chain.empty()) {
+    return absl::InvalidArgumentError("Certificate chain string is empty.");
+  }
+  BIO* cert_chain_bio = BIO_new_mem_buf(cert_chain.data(), cert_chain.size());
+  STACK_OF(X509_INFO)* cert_stack =
+      PEM_X509_INFO_read_bio(cert_chain_bio, nullptr, nullptr, nullptr);
+  int num_certs = sk_X509_INFO_num(cert_stack);
+  gpr_log(GPR_ERROR, "Num of certs: %d", num_certs);
+  bool is_null = false;
+  bool are_all_certs_invalid = num_certs == 0;
```

Comment:
We still have to free the OpenSSL resources before returning anything to avoid memory leaks or crashes. That's why the `are_all_certs_invalid` is used.

Reply:
I'm using smart pointers now so this doesn't apply anymore.

Reply:
The code looks much cleaner with smart pointers. Thank you!